confluent_kafka/avro/schema.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python
+#
+# Copyright 2018 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from avro.schema import PrimitiveSchema
+
+# Python 2 considers int an instance of str
+try:
+    string_type = basestring  # noqa
+except NameError:
+    string_type = str
+
+
+class GenericAvroRecord(dict):
+    """
+    Pairs an AvroRecord with it's schema
+
+    :param schema schema: A parsed Avro schema.
+    :param dict record: Wraps existing dict in GenericAvroRecord.
+    :raises ValueError: If schema is None.
+    :returns: Avro record with its schema.
+    :rtype: GenericAvroRecord
+    """
+    __slots__ = ['schema']
+
+    def __init__(self, schema, record=None):
+        if schema is None:
+            raise ValueError("schema must not be None")
+        self.schema = schema
+        if record is not None:
+            self.update(record)
+
+    def put(self, key, value):
+        self[key] = value
+
+
+def get_schema(datum):
+    if isinstance(datum, GenericAvroRecord):
+        return datum.schema
+    elif (datum is None):
+        return PrimitiveSchema("null")
+    elif isinstance(datum, string_type):
+        return PrimitiveSchema('string')
+    elif isinstance(datum, bool):
+        return PrimitiveSchema('boolean')
+    elif isinstance(datum, int):
+        return PrimitiveSchema('int')
+    elif isinstance(datum, float):
+        return PrimitiveSchema('float')
```

Comment:
python floats are double precision and if you don't use a `'double'` primitive schema here, serialization will be lossy in most cases which seems bad.

same with int - python seems to have unified int's to be mean long in more recent versions of the language, and I think just using a `'long'` primitive schema type for all int values is best.

Reply:
I did forget to add the double, I'll add that in. I do not think we should be silently promoting floats. The avro implementation actually ends up calling write_long within write_int anyway so there is nothing to be gained there. It's much more straightforward to say an int is an int. 

https://github.com/apache/avro/blob/master/lang/py/src/avro/io.py#L300-L304

Reply:
python doesn't have any data type that corresponds to avro's float or avro's int.

python float is equivalent to avro double.

pre-python3, I *think* python int's are platform dependent. On my machine they'e 64 bit (i.e. equivalent to avro long).

in python3, all integer values are arbitrary precision (i.e. there is no corresponding avro type, but the best choice is long, presumably the avro library has some error checking).